File: 11_ Testing unitario con Spring/spring-testing-unitario/src/main/java/com/example/service/BookServiceImpl.java
package com.example.service;

import com.example.model.Book;
import com.example.repository.BookRespository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class BookServiceImpl implements BookService {

    private BookRespository repository;


    public BookServiceImpl(BookRespository repository) {
        this.repository = repository;
    }

    @Override
    public List<Book> findAll() {
        return this.repository.findAll();
    }

    @Override
    public Optional<Book> findById(Long id) {
        return this.repository.findById(id);
    }
}

File: 20_ Introuduccion a Spring WebFluix con Kotlin/spring-webflux-kotlin/src/main/kotlin/com/example/service/BookServiceImpl.kt
package com.example.service

import com.example.domain.Book
import com.example.repository.BookRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Service
class BookServiceImpl(@Autowired val bookRepository: BookRepository) : BookService {

    private fun upperText(book: Book): Book{
        book.title = book.title?.uppercase()
        return book
    }

    override fun findById(id: Long): Mono<Book> {
        return bookRepository.findById(id).map { book -> upperText(book) }
    }

    override fun findAll(): Flux<Book> {
        return bookRepository.findAll().map(::upperText)
    }

    override fun create(book: Book): Mono<Book> {
        return if (book.id != null)
            Mono.error(IllegalArgumentException(" id a de ser nulo"))
        else
            bookRepository.save(book)
    }

    override fun update(book: Book): Mono<Book> {
        return bookRepository.existById(book.id!!).flatMap { exist ->
            if (exist) {
                return@flatMap bookRepository.save(book)
            } else {
                return@flatMap Mono.error(IllegalArgumentException("El libro tiene que existir"))
            }
        }
    }

    override fun deleteById(id: Long): Mono<Void> {
        return bookRepository.deleteById(id)
        }


    override fun deleteAll(): Mono<Void> {
        return bookRepository.deleteAll()
    }
}